drivers/gpio/gpio_mcp23xxx.c

```diff
@@ -0,0 +1,489 @@
+		LOG_ERR("Error setting pin pull up/pull down (%d)", ret);
+		goto done;
+	}
+
+done:
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23xxx_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = read_port_regs(dev, REG_GPIO, &buf);
+	if (ret == 0) {
+		*value = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23xxx_port_set_masked_raw(const struct device *dev, uint32_t mask, uint32_t value)
+{
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf = (buf & ~mask) | (mask & value);
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp23xxx_port_set_bits_raw(const struct device *dev, uint32_t mask)
+{
+	return mcp23xxx_port_set_masked_raw(dev, mask, mask);
+}
+
+static int mcp23xxx_port_clear_bits_raw(const struct device *dev, uint32_t mask)
+{
+	return mcp23xxx_port_set_masked_raw(dev, mask, 0);
+}
+
+static int mcp23xxx_port_toggle_bits(const struct device *dev, uint32_t mask)
+{
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf ^= mask;
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	return ret;
+}
+
+static int mcp23xxx_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+					    enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	return -ENOTSUP;
+}
+
+static const struct gpio_driver_api api_table = {
+	.pin_configure = mcp23xxx_pin_cfg,
+	.port_get_raw = mcp23xxx_port_get_raw,
+	.port_set_masked_raw = mcp23xxx_port_set_masked_raw,
+	.port_set_bits_raw = mcp23xxx_port_set_bits_raw,
+	.port_clear_bits_raw = mcp23xxx_port_clear_bits_raw,
+	.port_toggle_bits = mcp23xxx_port_toggle_bits,
+	.pin_interrupt_configure = mcp23xxx_pin_interrupt_configure,
+};
+
+/**
+ * @brief Initialization function of MCP230XX
+ *
+ * @param dev Device struct.
+ * @return 0 if successful. Otherwise <0 is returned.
+ */
+static int mcp23xxx_init(const struct device *dev)
+{
+	const struct mcp23xxx_config *config = dev->config;
+	struct mcp23xxx_drv_data *drv_data = (struct mcp23xxx_drv_data *)dev->data;
+
+	if (config->ngpios != 8 && config->ngpios != 16) {
+		LOG_ERR("Invalid value ngpios=%u. Expected 8 or 16!", config->ngpios);
+		return -EINVAL;
+	}
+
+	if (!config->bus_fn(dev)) {
+		return -ENODEV;
+	}
```

Comment:
`bus_fn` can return -ENODEV, and here just propagate it.

Reply:
Yeah, I waffled on this too. Refactored this to a version that returns an errno, and changed this to just propagate if getting negative err.